### ISSUE TITLE
Should S3267 raise for iterations over (non-generic) IEnumerable?

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#4972]
On the support forum an (FP has been raised)[https://community.sonarsource.com/t/51187]. I've responded there, but thought also to raise the issue/question here.

I think there are two ways to fix the issue:
 1. Suggest the change including a `.Cast<>()` before the `.Where()` and/or `.Select()`
 2. Skip the rule on non-generic `IEnumerable`

``` C#
static class S3267
{
    public static void IterateNonGeneric(NonGeneric items)
    {
        foreach (Item item in items)
        {
            if (item.Value > 66) Console.WriteLine(item.Value);
        }
    }
    public static void IterateGeneric(IEnumerable<Item> items)
    {
        foreach (Item item in items)
        {
            if (item.Value > 66) Console.WriteLine(item.Value);
        }
    }
}

public class Item
{
    public int Value { get; set; }
}

public class NonGeneric : IEnumerable
{
    public IEnumerator GetEnumerator() => Array.Empty<object>().GetEnumerator();
}
```

What do you think?

**Comment by costin-zaharia-sonarsource** [comment]
Hi @Corniel, thanks for reporting the issue. I agree that we should avoid raising the issue for non-generic `IEnumerable`.
I'm closing this one since we have #4996  for this.
